botocore/regions.py

```diff
@@ -466,27 +469,78 @@ class CredentialBuiltinResolver:
         self._account_id_endpoint_mode = account_id_endpoint_mode
         self._validate_account_id_endpoint_mode()
 
-    def resolve_account_id_builtin(self, builtin_configured, builtin_value):
-        """Resolve the ``AWS::Auth::AccountId`` builtin."""
-        acct_id_ep_mode = self._account_id_endpoint_mode
-        mode_disabled = acct_id_ep_mode == 'disabled'
-        no_credentials = self._credentials is None
-        if not builtin_configured or mode_disabled or no_credentials:
-            return None
+    def _validate_account_id_endpoint_mode(self):
+        valid_modes = self.VALID_ACCOUNT_ID_ENDPOINT_MODES
+        if self._account_id_endpoint_mode not in valid_modes:
+            error_msg = (
+                f'Invalid value "{self._account_id_endpoint_mode}" '
+                'for account_id_endpoint_mode. Valid values are: '
+                f'{", ".join(valid_modes)}.'
+            )
+            raise InvalidConfigError(error_msg=error_msg)
 
-        if builtin_value is not None:
-            return builtin_value
+    def resolve(self, param_definitions, builtins):
+        """Resolve endpoint builtins sourced from credentials."""
+        builtins_to_resolve = self._builtins_to_resolve(param_definitions)
+        if not builtins_to_resolve:
+            return
 
         frozen_creds = self._credentials.get_frozen_credentials()
-        account_id = frozen_creds.account_id
+        for builtin_name, builtin_key in builtins_to_resolve.items():
+            credential_value = getattr(frozen_creds, builtin_name)
+            builtin_value = builtins.get(builtin_key) or credential_value
+            builtins[builtin_key] = self._validate_builtin(
+                builtin_name, builtin_value
+            )
+
+    def _builtins_to_resolve(self, param_definitions):
+        builtins_to_resolve = {}
+        if self._credentials is None:
+            return builtins_to_resolve
+
+        for builtin_name, builtin_key in self.CREDENTIAL_BUILTINS.items():
+            if self._should_resolve_builtin(builtin_name, param_definitions):
+                builtins_to_resolve[builtin_name] = builtin_key
+
+        return builtins_to_resolve
+
+    def _should_resolve_builtin(self, builtin_name, param_definitions):
+        should_resolve = getattr(self, f'_should_resolve_{builtin_name}')
+        return should_resolve(param_definitions)
+
+    def _should_resolve_account_id(self, param_definitions):
+        has_builtin = self._builtin_configured(param_definitions, 'AccountId')
+        mode_enabled = self._account_id_endpoint_mode != 'disabled'
+        return has_builtin and mode_enabled
+
+    def _should_resolve_scope(self, param_definitions):
+        return self._builtin_configured(param_definitions, 'CredentialScope')
+
+    def _builtin_configured(self, param_definitions, param_name):
+        param_def = param_definitions.get(param_name)
+        return param_def is not None and param_def.builtin is not None
```

Comment:
We can. I think I opted to keep it because it is similar syntactically to the `_validate_builtin` logic and avoids the use of conditionals. Additionally we will need to update the mapping to use the pascal case `AccountId` and `CredentialScope`. I'll have something up soon.